File: app/src/main/java/org/ole/planet/myplanet/model/RealmAnswer.java
package org.ole.planet.myplanet.model;

import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.utilities.Utilities;

import java.util.HashMap;

import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class RealmAnswer extends RealmObject {
    @PrimaryKey
    private String id;
    private String value;
    private RealmList<String> valueChoices;
    private int mistakes = 0;
    private boolean passed;
    private int grade;
    private String examId;
    private String questionId;
    private String submissionId;

    public static JsonArray serializeRealmAnswer(RealmList<RealmAnswer> answers) {
        Utilities.log("Ans size " + answers.size());
        JsonArray array = new JsonArray();
        for (RealmAnswer ans : answers) {
            array.add(createObject(ans));
        }
        return array;
    }

    private static JsonObject createObject(RealmAnswer ans) {
        JsonObject object = new JsonObject();
        if (!TextUtils.isEmpty(ans.getValue())) {
            object.addProperty("value", ans.getValue());
        } else {
            object.add("value", ans.getValueChoicesArray());
        }
        object.addProperty("mistakes", ans.getMistakes());
        object.addProperty("passed", ans.isPassed());
        return object;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public JsonArray getValueChoicesArray() {
        JsonArray array = new JsonArray();
        if (valueChoices == null) {
            return array;
        }
        for (String choice : valueChoices) {
            array.add(new Gson().fromJson(choice, JsonObject.class));
        }
        return array;
    }

    public RealmList<String> getValueChoices() {
        return valueChoices;
    }

    public void setValueChoices(HashMap<String, String> map, boolean isLastAnsvalid) {
        if (!isLastAnsvalid)
            this.valueChoices.clear();
        for (String key : map.keySet()) {
            JsonObject ob = new JsonObject();
            ob.addProperty("id", map.get(key));
            ob.addProperty("text", key);
            this.valueChoices.add(new Gson().toJson(ob));
        }
    }

    public void setValueChoices(RealmList<String> valueChoices) {
        this.valueChoices = valueChoices;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }

    public int getMistakes() {
        return mistakes;
    }

    public void setMistakes(int mistakes) {
        this.mistakes = mistakes;
    }

    public boolean isPassed() {
        return passed;
    }

    public void setPassed(boolean passed) {
        this.passed = passed;
    }

    public int getGrade() {
        return grade;
    }

    public void setGrade(int grade) {
        this.grade = grade;
    }

    public String getExamId() {
        return examId;
    }

    public void setExamId(String examId) {
        this.examId = examId;
    }

    public String getQuestionId() {
        return questionId;
    }

    public void setQuestionId(String questionId) {
        this.questionId = questionId;
    }

    public String getSubmissionId() {
        return submissionId;
    }

    public void setSubmissionId(String submissionId) {
        this.submissionId = submissionId;
    }
}

File: app/src/main/java/org/ole/planet/myplanet/model/RealmAnswer.kt
package org.ole.planet.myplanet.model

import android.text.TextUtils
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import io.realm.kotlin.ext.realmListOf
import io.realm.kotlin.types.RealmList
import io.realm.kotlin.types.RealmObject
import io.realm.kotlin.types.annotations.PrimaryKey

class RealmAnswer : RealmObject {
    @PrimaryKey
    var id: String? = null
    var value: String? = null
    var valueChoices: RealmList<String> = realmListOf()
    var mistakes: Int = 0
    var isPassed: Boolean = false
    var grade: Int = 0
    var examId: String? = null
    var questionId: String? = null
    var submissionId: String? = null

    val valueChoicesArray: JsonArray get() {
        val array = JsonArray()
        for (choice in valueChoices) {
            array.add(Gson().fromJson(choice, JsonObject::class.java))
        }
        return array
    }

    fun setValueChoices(map: HashMap<String, String>?, isLastAnsValid: Boolean) {
        if (!isLastAnsValid) {
            valueChoices.clear()
        }
        map?.forEach { (key, value) ->
            val ob = JsonObject().apply {
                addProperty("id", value)
                addProperty("text", key)
            }
            valueChoices.add(Gson().toJson(ob))
        }
    }

    companion object {
        @JvmStatic
        fun serializeRealmAnswer(answers: RealmList<RealmAnswer>): JsonArray {
            val array = JsonArray()
            answers.forEach { ans ->
                array.add(createObject(ans))
            }
            return array
        }

        private fun createObject(ans: RealmAnswer): JsonObject {
            val jsonObject = JsonObject()
            if (!TextUtils.isEmpty(ans.value)) {
                jsonObject.addProperty("value", ans.value)
            } else {
                jsonObject.add("value", ans.valueChoicesArray)
            }
            jsonObject.addProperty("mistakes", ans.mistakes)
            jsonObject.addProperty("passed", ans.isPassed)
            return jsonObject
        }
    }
}
